Add a config converter for hyphenated enums

machi1990 opened the pull request:
Fixes #2678 

/cc @dmlloyd

emmanuelbernard commented:
Ping @dmlloyd

dmlloyd reviewed (commented):
Here is what I'm presently struggling with on this change.

* Hyphenation is very specific to enum config values.  Having an annotation that only works in one very narrow situation seems wrong somehow.
* Some enum values - as you rightly pointed out - should not be converted using this feature.
* This is not the first (and will not be the last) case where conversion behavior has to be configured; I think it might be better to have a general-purpose conversion mechanism.
* But which should be the default behavior for enums?  If the default is to pass them "as-is", making hyphenation be an exception to the rule, then we'll probably end up with a mix of things using each approach.  But if the default is to always hyphenate, it could lead to some weird or surprising problems in some cases.
* This *should* be possible to do using `Converters`.  I think maybe what we really should have is a general `@ConvertWith` annotation which accepts a `Converter` class object.  If the annotation is present, we override the converter in use for that item.  If it is absent, we use the default conversion for that item.
* If we decide that hyphenation *should* be the default for enums, then the way we should handle that would be to check if the config value type is an enum, and if so, default to using our special converter which does basically what your current `ConfigUtils.convertToEnum` is doing right now, just wrapped with the `Converter` interface.  If the user specifies a `@ConvertWith` annotation then we should use the one specified.  We should also provide a converter class which the user can specify to get the classic enum behavior - or perhaps use a special marker type to indicate that (e.g. a `DefaultConverter.class` marker interface).

Does that make sense?  Sorry for asking you to change everything again, but for me sometimes problems or ideas only become visible once I'm looking at code.

machi1990 commented on core/runtime/src/main/java/io/quarkus/runtime/annotations/Hyphenate.java:
```diff
@@ -0,0 +1,21 @@
+    /**
+     * The value indicating whether to activate enum conversion of hyphenated values.
+     */
+    boolean value() default false;
```
Yeah. I am more and more inclined to giving `@ConvertWith` a go. :-)

machi1990 commented:
@dmlloyd thanks for the feedback, highly appreciated. 

> * Hyphenation is very specific to enum config values.  Having an annotation that only works in one very narrow situation seems wrong somehow.
> * Some enum values - as you rightly pointed out - should not be converted using this feature.
> * This is not the first (and will not be the last) case where conversion behavior has to be configured; I think it might be better to have a general-purpose conversion mechanism.

I do agree with the above points. You are making a good case and I think having as general purpose API would make things easer later on for others conversion that need to be configured. 

> * This _should_ be possible to do using `Converters`.  I think maybe what we really should have is a general `@ConvertWith` annotation which accepts a `Converter` class object.  If the annotation is present, we override the converter in use for that item.  If it is absent, we use the default conversion for that item. 

Is there a way to override the converter easier? While doing my current approach, I had to first convert to `String` and then apply a specific converter i.e `ConfigUtils.convertToEnum` to get the desirable value. I am not a fan of the approach but I could not find better. What did you make of it? 

> * If we decide that hyphenation _should_ be the default for enums, then the way we should handle that would be to check if the config value type is an enum, and if so, default to using our special converter which does basically what your current `ConfigUtils.convertToEnum` is doing right now, just wrapped with the `Converter` interface. 

From what I have seen, we have a mixer of the two: hyphenation e.g `TransactionIsolationLevel` and default enum's `name` e.g  file `OverflowAction` . Opting for a default will surely cause some breaking change, is that affordable ?. 
Will having an hyphenation converter that can be used with `@ConvertWith` on a case by case basis, be an option?  

> If the user specifies a `@ConvertWith` annotation then we should use the one specified.  We should also provide a converter class which the user can specify to get the classic enum behavior - or perhaps use a special marker type to indicate that (e.g. a `DefaultConverter.class` marker interface).

I do not see the value of `DefaultConverter.class`. I would generally say, if `@ConvertWith` annotation is not used, then apply normal conversion behaviour: based in BuiltIn converters.

dmlloyd commented:
> > * This _should_ be possible to do using `Converters`.  I think maybe what we really should have is a general `@ConvertWith` annotation which accepts a `Converter` class object.  If the annotation is present, we override the converter in use for that item.  If it is absent, we use the default conversion for that item.
> 
> Is there a way to override the converter easier? While doing my current approach, I had to first convert to `String` and then apply a specific converter i.e `ConfigUtils.convertToEnum` to get the desirable value. I am not a fan of the approach but I could not find better. What did you make of it?

Unfortunately not.  I think it might be a good idea to enhance SmallRye Config's `SmallRyeConfig` class to have a method like this:

```java
public <T> T getValue(String name, Converter<T> customConverter) {
    // ...
}
```

And possibly also add variants for the `getValues` and `getOptionalValue` methods as well.  It might be something that would be nice to push up into the specification in fact.

But for now, just reading the value as a `String` is the best choice that I can see.

> > * If we decide that hyphenation _should_ be the default for enums, then the way we should handle that would be to check if the config value type is an enum, and if so, default to using our special converter which does basically what your current `ConfigUtils.convertToEnum` is doing right now, just wrapped with the `Converter` interface.
> 
> From what I have seen, we have a mixer of the two: hyphenation e.g `TransactionIsolationLevel` and default enum's `name` e.g file `OverflowAction` . Opting for a default will surely cause some breaking change, is that affordable ?.
> Will having an hyphenation converter that can be used with `@ConvertWith` on a case by case basis, be an option?

That is my worry.  There are two cases:

* If we cause hyphenation to be the default, then we should update all the points in the current code base which use enums but which should not use hyphenation.  If there are cases outside of the main Quarkus repository, it's possible that those extensions might start accepting configuration values which were previously disallowed.  I wonder if that is really a serious problem?
* If we do not cause hyphenation to be the default, then we should do the opposite and find all the values which can be hyphenated.  But since it is not the default, will we eventually have a random mix of behaviors where some values are hyphenated and some are not, for no good reason (other than that the author did not know what was possible)?

I'm thinking that the second case is the worse case, but I'm not 100% certain of that yet.

> > If the user specifies a `@ConvertWith` annotation then we should use the one specified.  We should also provide a converter class which the user can specify to get the classic enum behavior - or perhaps use a special marker type to indicate that (e.g. a `DefaultConverter.class` marker interface).
> 
> I do not see the value of `DefaultConverter.class`. I would generally say, if `@ConvertWith` annotation is not used, then apply normal conversion behaviour: based in BuiltIn converters.

I agree mostly, however if we make hyphenation be the default, then the user would have no way to explicitly say "use non-hyphenation" because that converter is provided by SmallRye Config as an implicit converter, which do not have public classes that could be referenced from the annotation.  It's also possible that this same problem could occur in other cases if we decide to introduce other "enhanced" converters for types that are normally covered by implicit converters.  So it seems to me that we do need *some* solution or way of saying "convert this with the default converter, whatever that may be".  Does that make sense?

machi1990 commented:
> Unfortunately not. I think it might be a good idea to enhance SmallRye Config's `SmallRyeConfig` class to have a method like this:
> 
> ```java
> public <T> T getValue(String name, Converter<T> customConverter) {
>     // ...
> }
> ```
> 
> And possibly also add variants for the `getValues` and `getOptionalValue` methods as well. It might be something that would be nice to push up into the specification in fact.
> 
> But for now, just reading the value as a `String` is the best choice that I can see.

Okay, enhancing a class seems a long shot for me so  I'll stick with the current option for now. 
 
> That is my worry. There are two cases:
> 
> * If we cause hyphenation to be the default, then we should update all the points in the current code base which use enums but which should not use hyphenation.  If there are cases outside of the main Quarkus repository, it's possible that those extensions might start accepting configuration values which were previously disallowed.  I wonder if that is really a serious problem?
> * If we do not cause hyphenation to be the default, then we should do the opposite and find all the values which can be hyphenated.  But since it is not the default, will we eventually have a random mix of behaviors where some values are hyphenated and some are not, for no good reason (other than that the author did not know what was possible)?
> 

> I'm thinking that the second case is the worse case, but I'm not 100% certain of that yet.

Maybe we have to go with the lesser evil :-), that is the first option. And probably add a documentation in extension guide about the change and how to handle it with the `DefaultConverter.class`. 
On what concerns Quarkus Repository, the `Protocol` enum is a good  candidate for opting out to a DefaultConverter class, it can be used as an example: see below.
 
> > > If the user specifies a `@ConvertWith` annotation then we should use the one specified.  We should also provide a converter class which the user can specify to get the classic enum behavior - or perhaps use a special marker type to indicate that (e.g. a `DefaultConverter.class` marker interface).
> > 
> > 
> > I do not see the value of `DefaultConverter.class`. I would generally say, if `@ConvertWith` annotation is not used, then apply normal conversion behaviour: based in BuiltIn converters.
> 
> I agree mostly, however if we make hyphenation be the default, then the user would have no way to explicitly say "use non-hyphenation" because that converter is provided by SmallRye Config as an implicit converter, which do not have public classes that could be referenced from the annotation. It's also possible that this same problem could occur in other cases if we decide to introduce other "enhanced" converters for types that are normally covered by implicit converters. So it seems to me that we do need _some_ solution or way of saying "convert this with the default converter, whatever that may be". Does that make sense?

Taking `Protocol` enum for example:
1. 
```
     /**
     * The list of protocols to explicitly enable.
     */
    @DefaultConvter 
    @ConfigItem(defaultValue = "TLSv1.3,TLSv1.2")
    public List<Protocol> protocols;
```

or 

2.
```
     /**
     * The list of protocols to explicitly enable.
     */
    @ConvertWith(DefaultConvert.class)
    @ConfigItem(defaultValue = "TLSv1.3,TLSv1.2")
    public List<Protocol> protocols;
```

3. ??

dmlloyd commented:
> ```
>      /**
>      * The list of protocols to explicitly enable.
>      */
>     @DefaultConverter 
>     @ConfigItem(defaultValue = "TLSv1.3,TLSv1.2")
>     public List<Protocol> protocols;
> ```

I like this option.  You're right, it does look better than a marker type.  We'll have to check to make sure that both annotations aren't given, however.

machi1990 commented:
> > ```
> >      /**
> >      * The list of protocols to explicitly enable.
> >      */
> >     @DefaultConverter 
> >     @ConfigItem(defaultValue = "TLSv1.3,TLSv1.2")
> >     public List<Protocol> protocols;
> > ```
> 
> I like this option. You're right, it does look better than a marker type. We'll have to check to make sure that both annotations aren't given, however.

Nice. I'll give this a re-spin tomorrow. Thanks @dmlloyd.

machi1990 committed:
feat(configuration): add ConvertWith and DefaultConverter annotations.

Adds a default enum converter for hyphenated values to override implicit enum default converter.
Fixes #2678

machi1990 committed:
feat(configuration): take into account converters provided through ConvertWith annotation

If both ConvertWith and DefaultConverter annotation are provided throws an error. Make conversion
from hyphenated values a default for enums

dmlloyd commented on core/deployment/src/main/java/io/quarkus/deployment/configuration/BooleanConfigType.java:
```diff
@@ -50,23 +53,24 @@ public void generateAcceptConfigurationValue(final BytecodeCreator body, final R
         try {
-            field.setBoolean(enclosing, config.getOptionalValue(name.toString(), Boolean.class)
-                    .orElse(Boolean.FALSE).booleanValue());
+            Optional<Boolean> optionalValue = (Optional<Boolean>) ConfigUtils.getOptionalValue(config, name.toString(),
```
This is an example of a cast that would be unnecessary by generifying `ConfigUtils#get*Value` (see other comment on that class).

dmlloyd commented on core/deployment/src/main/java/io/quarkus/deployment/configuration/ConfigDefinition.java:
```diff
@@ -264,59 +274,113 @@ private GroupConfigType processConfigGroup(final String containingName, final Co
-                        accessorFinder, javadocKey));
+
+                Type mapValueType = typeOfParameter(fieldType, 1);
+                Class<?> mapValueRawType = mapValueType instanceof ParameterizedType ? rawTypeOf(mapValueType)
```
You can just make this be `Class<?> mapValueRawType = rawTypeOf(mapValueType);` and it should work fine.

dmlloyd commented on core/deployment/src/main/java/io/quarkus/deployment/configuration/ConfigDefinition.java:
```diff
@@ -264,59 +274,113 @@ private GroupConfigType processConfigGroup(final String containingName, final Co
+
+        if (convertWith != null) {
+            if (clazz.isEnum()) {
+                return HyphenateEnumConverter.class;
```
Doesn't this mean that the converter class specified with `ConvertWith` will be ignored for `enum` values?  I would think that we'd still want the user to be able to override the converter, so the behavior would be like this:

| Member has | Converter used is |
| -- | -- |
| nothing | `HyphenatedEnumConverter` |
| `@DefaultConverter` | the standard SmallRye converter |
| `@ConvertWith(Xxx.class)` | the converter class `Xxx` |

machi1990 replied:
Good catch.

dmlloyd commented on core/deployment/src/main/java/io/quarkus/deployment/configuration/ConfigDefinition.java:
```diff
@@ -264,59 +274,113 @@ private GroupConfigType processConfigGroup(final String containingName, final Co
+                try {
+                    final Method method = converterClass.getMethod("convert", String.class);
+                    final Type type = method.getAnnotatedReturnType().getType();
+                    if (clazz.equals(type)) {
```
This should be `class.isAssignableFrom(getRawTypeOf(type))` I would think, because the converter output type can be more specific than the accepting member's type (for example I could use a `Converter<String>` to assign a value into a field of type `Object`).

dmlloyd commented on core/deployment/src/main/java/io/quarkus/deployment/configuration/FloatConfigType.java:
```diff
@@ -54,8 +56,8 @@ public void generateAcceptConfigurationValue(final BytecodeCreator body, final R
         try {
-            final Float floatValue = config.getValue(name.toString(), Float.class);
-            final float f = floatValue != null ? floatValue.floatValue() : 0f;
+            final Object floatValue = ConfigUtils.getValue(config, name.toString(), Float.class, converterClass);
```
Another point where making `ConfigUtils.getValue` be generic will clean up the code.

dmlloyd reviewed (changes requested):
This is looking a lot better.  The remaining changes are all relatively small I think.  Thanks!

machi1990 commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java:
```diff
@@ -20,18 +29,116 @@ private ConfigUtils() {
+     * @param converterClass - The converter class to use
+     * @return the value in appropriate type
+     */
+    public static Object getValue(SmallRyeConfig config, String configName, Class<?> objectType,
```
I'll clean up the unsafe casts. Thanks for spotting this.

machi1990 commented on core/runtime/src/main/java/io/quarkus/runtime/logging/AsyncConfig.java:
```diff
@@ -22,6 +23,7 @@
     /**
      * Determine whether to block the publisher (rather than drop the message) when the queue is full
      */
+    @DefaultConverter
```
It's one of those breaking changes we talked about. At this stage some users are already using `BLOCK`.

dmlloyd replied:
If we hyphenate the enum constants, and we hyphenate the user input, then they could continue to have `BLOCK` as a value without breaking compatibility.

machi1990 replied:
> It's one of those breaking changes we talked about. At this stage some users are already using `BLOCK`.

We have a fallback `Enum.valueOf(type, value)` option in `HyphenatedEnumConverter`. So I think this could work too.

dmlloyd replied:
In that case we'd only support the fully hyphenated version (`block`) or the exact original enum constant (`BLOCK`).  Hyphenating both would allow supporting any value that maps to the hyphenated version (such as `Block`), which is slightly more flexible.

machi1990 replied:
Good remark on flexibility. I'll update this and the corresponding tests too.

machi1990 commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java:
```diff
@@ -20,18 +29,116 @@ private ConfigUtils() {
+        }
+
+        try {
+            return converterClass.newInstance();
```
> I'm not really happy about having to use reflection at runtime. 

It's among the points that I wanted to see with you to get better idea than using reflection at runtime. So thanks for spotting this too.

> Is there a way that we could maybe populate a map of class to converter instance during static init? 

Will having a static `ConfigUtils#setConverterFor(type, converterClass)` called from `ConfigDefinition#loadEnhancedConverter()` be a reasonable approach? . The method will then cache the converter in a map .

dmlloyd replied:
Since we need the converters to all be instantiated during static init, we'd have to store them somewhere during `io.quarkus.deployment.steps.ConfigurationSetup#finalizeConfigLoader`.  Lazily caching won't be good enough for this case.

Also we have to be careful only to cache converters that are used for config items that are visible at run time (`ConfigPhase#RUN_TIME` and `ConfigPhase#BUILD_AND_RUN_TIME_FIXED`).

machi1990 replied:
> Since we need the converters to all be instantiated during static init, we'd have to store them somewhere during `io.quarkus.deployment.steps.ConfigurationSetup#finalizeConfigLoader`. Lazily caching won't be good enough for this case.

Thanks for the pointer. I'll check the method. 

> 
> Also we have to be careful only to cache converters that are used for config items that are visible at run time (`ConfigPhase#RUN_TIME` and `ConfigPhase#BUILD_AND_RUN_TIME_FIXED`).

+1

machi1990 replied:
The strategy opted is to cache runtime explicit converters to get rid of runtime instance creation using reflection. 

The code code is untouched for build time explicit converters as they are not cached.

@dmlloyd WDYT?

machi1990 commented on core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java:
```diff
@@ -499,6 +508,26 @@ public void write(final String name, final byte[] data) {
                 // custom run time converters
                 carc.invokeStaticMethod(CS_POPULATE_CONVERTERS, builder);
 
+                // cache explicit converts and make them available during runtime
```
@dmlloyd I opted for this approach. WDYT?

dmlloyd replied:
This looks better, thanks!

machi1990 replied:
Thanks for the initial suggestion :-)

machi1990 commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java:
```diff
@@ -20,18 +29,116 @@ private ConfigUtils() {
+        return converter.convert(rawValue);
+    }
+
+    private static Converter<?> getConverterOfType(Class<?> type, Class<? extends Converter<?>> converterClass) {
```
Done

dmlloyd commented:
I will be able to review this again on Monday.

dmlloyd commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java:
```diff
@@ -20,18 +30,161 @@ private ConfigUtils() {
      */
-    public static <T, C extends Collection<T>> C getDefaults(SmallRyeConfig config, String defaultValue, Class<T> itemType,
+    public static <T, C extends Collection<T>> C getDefaults(SmallRyeConfig config, String defaultValue, Class<?> itemType,
+            Class<Converter<?>> converterClass,
```
`Class<? extends Converter<T>>`

dmlloyd commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java:
```diff
@@ -20,18 +30,161 @@ private ConfigUtils() {
      * @return the collection (not {@code null})
      */
-    public static <T, C extends Collection<T>> C getDefaults(SmallRyeConfig config, String defaultValue, Class<T> itemType,
+    public static <T, C extends Collection<T>> C getDefaults(SmallRyeConfig config, String defaultValue, Class<?> itemType,
```
`itemType` should remain `Class<T>`.

dmlloyd commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java:
```diff
@@ -20,18 +30,161 @@ private ConfigUtils() {
+     * @return the values in appropriate type
+     */
+    public static ArrayList<?> getValues(SmallRyeConfig config, String configName, Class<?> objectType,
+            Class<Converter<?>> converterClass) {
```
`...ArrayList<T> getValues(..., Class<T> objectType, Class<? extends Converter<T>> converterClass) {`

dmlloyd commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java:
```diff
@@ -20,18 +30,161 @@ private ConfigUtils() {
+     * @return the value
+     */
+    public static <T> T convert(SmallRyeConfig config, String value, Class<?> objectType,
+            Class<Converter<T>> converterClass) {
```
`...Class<T> objectType, Class<? extends Converter<T>> converterClass) {`

dmlloyd commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java:
```diff
@@ -20,18 +30,161 @@ private ConfigUtils() {
+        return converter.convert(rawValue);
+    }
+
+    private static Converter<?> getConverterOfType(Class<?> type, Class<Converter<?>> converterType) {
```
`Converter<T> getConverterOfType(Class<T> type, Class<? extends Converter<T>> converterType) {`

dmlloyd commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java:
```diff
@@ -20,18 +30,161 @@ private ConfigUtils() {
+    }
+
+    private static Converter<?> getConverterOfType(Class<?> type, Class<Converter<?>> converterType) {
+        Converter<?> converter = EXPLICIT_RUNTIME_CONVERTERS_CACHE.get(new ConverterClassHolder(type, converterType));
```
`@SuppressWarnings("unchecked") Coverter<T> converter = (Converter<T>) ...`

dmlloyd commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java:
```diff
@@ -20,18 +30,161 @@ private ConfigUtils() {
+            return (T) config.convert(value, objectType);
+        }
+
+        final Converter<T> converter = getConverterOfType(objectType, (Class) converterClass);
```
Raw cast.

dmlloyd commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java:
```diff
@@ -20,18 +30,161 @@ private ConfigUtils() {
+    public static <T> T convert(SmallRyeConfig config, String value, Class<?> objectType,
+            Class<Converter<T>> converterClass) {
+        if (converterClass == null) {
+            return (T) config.convert(value, objectType);
```
This should just be `config.convert(value, objectType)` once `objectType` is fixed.

dmlloyd commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java:
```diff
@@ -20,18 +30,161 @@ private ConfigUtils() {
+            return config.getOptionalValue(configName, objectType);
+        }
+
+        final Converter<T> converter = getConverterOfType(objectType, (Class) converterClass);
```
Raw cast.

dmlloyd commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java:
```diff
@@ -20,18 +30,161 @@ private ConfigUtils() {
+            return config.getValue(configName, objectType);
+        }
+
+        final Converter<T> converter = getConverterOfType(objectType, (Class) converterClass);
```
Raw cast

dmlloyd commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java:
```diff
@@ -20,18 +30,161 @@ private ConfigUtils() {
         for (String item : items) {
-            collection.add(config.convert(item, itemType));
+            if (converterClass == null) {
+                collection.add((T) config.convert(item, itemType));
```
With generics changed as described above, this cast should not be needed.

dmlloyd reviewed (changes requested):
Thanks, this is much closer to what we want!  There are a few generics changes that are needed but I think this approach is correct overall.

machi1990 committed:
refactor: applies the default hyphenated enum conversion.

machi1990 commented:
> Thanks, this is much closer to what we want! There are a few generics changes that are needed but I think this approach is correct overall.

@dmlloyd thanks for the suggestions. I have incorporated them. 
Native build timed out causing a failure. Is there away to restart it other than force pushing?

aloubyansky commented:
@dmlloyd could you please approve this one? Thanks.

dmlloyd commented:
It still needs a few more fixes, but I haven't gotten around to another round of review.

dmlloyd committed:
Misc. generics fixups

dmlloyd commented:
I've added a commit with the remaining generics fixes.  It's getting a bit ugly but once the new SmallRye Config is integrated, I should be able to do a massive, massive cleanup with a tremendous reduction in complexity and code size.

machi1990 commented:
> I've added a commit with the remaining generics fixes. It's getting a bit ugly but once the new SmallRye Config is integrated, I should be able to do a massive, massive cleanup with a tremendous reduction in complexity and code size.

Cool, It is better now. When I was cleaning them, they were getting out of hand at some point. Thanks for the fix. 

We have a slight conflict on the doc. I'll go ahead and resolve it.

gsmet commented:
Merging as David marked it as ready. Thanks @machi1990 and @dmlloyd .
